Firebase/Database/Core/FServerValues.m

```diff
@@ -20,37 +20,102 @@
+    return @{kTimestamp : [NSNumber numberWithLongLong:millis]};
 }
 
 + (id)resolveDeferredValue:(id)val
-          withServerValues:(NSDictionary *)serverValues {
-    if ([val isKindOfClass:[NSDictionary class]]) {
-        NSDictionary *dict = val;
-        if (dict[kServerValueSubKey] != nil) {
-            NSString *serverValueType = [dict objectForKey:kServerValueSubKey];
-            if (serverValues[serverValueType] != nil) {
-                return [serverValues objectForKey:serverValueType];
-            } else {
-                // TODO: Throw unrecognizedServerValue error here
-            }
-        }
+              withExisting:(id<FNode>)existing
+              serverValues:(NSDictionary *)serverValues {
+    if (![val isKindOfClass:[NSDictionary class]]) {
+        return val;
+    }
+    NSDictionary *dict = val;
+    id op = dict[kServerValueSubKey];
+
+    if (op == nil) {
+        return val;
+    } else if ([op isKindOfClass:NSString.class]) {
+        return [FServerValues resolveScalarServerOp:op
+                                   withServerValues:serverValues];
+    } else if ([op isKindOfClass:NSDictionary.class]) {
+        return [FServerValues resolveComplexServerOp:op
+                                        withExisting:existing
+                                        serverValues:serverValues];
     }
     return val;
 }
 
++ (id)resolveScalarServerOp:(NSString *)op
+           withServerValues:(NSDictionary *)serverValues {
+    return serverValues[op];
+}
+
++ (id)resolveComplexServerOp:(NSDictionary *)op
+                withExisting:(id<FNode>)existing
+                serverValues:(NSDictionary *)serverValues {
+    // Only increment is supported as of now
+    if (op[kIncrement] == nil) {
+        return nil;
+    }
+
+    NSNumber *delta = op[kIncrement];
+    if (![existing isLeafNode]) {
+        return delta;
+    }
+    FLeafNode *existingLeaf = existing;
+    if (![existingLeaf.value isKindOfClass:NSNumber.class]) {
+        return delta;
+    }
+
+    BOOL incrLong = strcmp(delta.objCType, @encode(int)) == 0;
+    incrLong = incrLong || strcmp(delta.objCType, @encode(long)) == 0;
+
+    NSNumber *existingNum = existingLeaf.value;
+    BOOL baseLong = strcmp(existingNum.objCType, @encode(int)) == 0;
+    baseLong = baseLong || strcmp(existingNum.objCType, @encode(long)) == 0;
+
+    double doubleSum = delta.doubleValue + existingNum.doubleValue;
+    // Try to avoid coercing into double. Both sides mut be whole numbers that
+    // do not over/underflow.
+    if (incrLong && baseLong) {
+        long longSum = delta.longValue + existingNum.longValue;
+        // check for under/overflow
+        if (delta.longValue < 0 && existingNum.longValue < 0 && longSum > 0) {
+            return @(doubleSum);
+        }
+        if (delta.longValue > 0 && existingNum.longValue > 0 && longSum < 0) {
+            return @(doubleSum);
+        }
+        return @(doubleSum);
```

Comment:
If I am not mistaken, this should be `@(longSum)`. Can you write a test that verified the output type?